contracts/SeriesIndexer.sol

```diff
@@ -0,0 +1,448 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol" as Ownable;
+
+contract SeriesIndexer is Ownable.Ownable {
+    // Counter
+    uint256 private nextSeriesID = 1;
+    uint256 private nextArtistID = 1;
+
+    struct Series {
+        string metadata;
+        string contractTokenData;
+        uint256[] artistIDs;
+    }
+
+    // Artist Management
+    mapping(uint256 => address) private artistIDToAddress;
+    mapping(address => uint256) private addressToArtistID;
+    mapping(uint256 => bool) private ownerRightsRevokedForArtistID;
+
+    // Mappings for series and artists
+    mapping(uint256 => Series) private seriesDetails;
+    mapping(uint256 => uint256[]) private artistIDSeriesIDs;
+    mapping(uint256 => mapping(uint256 => bool)) private isArtistIDInSeries;// SeriesID => ArtistID => Bool
+    mapping(uint256 => mapping(uint256 => bool)) private seriesPendingCoArtist;// SeriesID => ArtistID => Bool
+
+    // Pending Requests Management
+    mapping(uint256 => uint256[]) private artistPendingCoArtistRequests;// artistID => SeriesID[]
+    mapping(uint256 => mapping(uint256 => uint256)) private artistPendingCoArtistRequestIndex;// artistID => SeriesID => Index
+
+    // Events
+    event ArtistAddressUpdated(uint256 indexed artistID, address indexed oldAddress, address indexed newAddress);
+    event SeriesIndexed(
+        uint256 indexed seriesID,
+        uint256[] artistIDs,
+        string metadata,
+        string seriesContractTokenCID
+    );
+    event SeriesUpdated(
+        uint256 indexed seriesID,
+        uint256[] artistIDs,
+        string metadata,
+        string seriesContractTokenCID
+    );
+    event SeriesDeleted(uint256 indexed seriesID);
+    event CoArtistProposed(uint256 indexed seriesID, uint256 indexed proposedArtistID);
+    event CoArtistConfirmed(uint256 indexed seriesID, uint256 indexed confirmedArtistID);
+    event CoArtistProposalCancelled(
+        uint256 indexed seriesID, 
+        uint256 indexed proposerArtistID, 
+        uint256 indexed cancelledArtistID
+    );
+
+    // Modifiers
+    modifier onlyOwnerOrArtist(uint256 seriesID) {
+        _checkOwnerOrArtist(seriesID);
+        _;
+    }
+
+    modifier onlyArtist(uint256 seriesID) {
+        _checkArtist(seriesID);
+        _;
+    }
+
+    modifier seriesExists(uint256 seriesID) {
+        require(bytes(seriesDetails[seriesID].metadata).length > 0, "Series does not exist");
+        _;
+    }
+    // ------------------------
+    // Internal Helper Functions
+    // ------------------------
+
+    function _checkOwnerOrArtist(uint256 seriesID) internal view {
```

Comment:
Should pass the address instead of assume the `msg.sender`. The caller should determine which one should be passed, that will be more extensible and readable. If we just want to stick with the `msg.sender`, we might need a better name like `_checkMsgSenderIsOwnerOrArtist`